lib/Twig/Template.php

```diff
@@ -270,29 +270,40 @@ public function renderBlock($name, array $context, array $blocks = array(), $use
     }
 
     /**
-     * Returns whether a block exists or not.
-     *
-     * This method is for internal use only and should never be called
-     * directly.
-     *
-     * This method does only return blocks defined in the current template
-     * or defined in "used" traits.
+     * Returns whether a block exists or not in the current context of the template.
      *
-     * It does not return blocks from parent templates as the parent
-     * template name can be dynamic, which is only known based on the
-     * current context.
+     * This method checks blocks defined in the current template
+     * or defined in "used" traits or defined in parent templates.
      *
-     * @param string $name The block name
+     * @param string $name    The block name
+     * @param array  $context The context
+     * @param array  $blocks  The current set of blocks
      *
      * @return bool true if the block exists, false otherwise
      *
-     * @see blockExists
-     *
      * @internal
      */
-    public function hasBlock($name)
+    public function hasBlock($name, array $context = null, array $blocks = array())
     {
-        return isset($this->blocks[(string) $name]);
+        if (null === $context) {
+            @trigger_error('The '.__METHOD__.' method is internal and should never be called; calling it directly is deprecated since version 1.28 and won\'t be possible anymore in 2.0.', E_USER_DEPRECATED);
```

Comment:
the deprecation warning is weird, as the method won't be removed entirely, but only this signature.